### ISSUE TITLE
Error 401 on fetching groupBy categories

### ISSUE THREAD

**Issue by m-radzikowski** [pbauerochse/youtrack-worklog-viewer#22]
I'm getting 401 unauthorized error. In settings I provided youtrack URL (http://youtrack/ - in our company VPN) and REST API user and password. I'm admin (role Global System Admin) so permissions souldn't be an issue here.

Youtrack version 7.0 build 29566.

```
09:43:44.882 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Showing settings dialogue
09:43:44.958 [JavaFX Application Thread] DEBUG d.p.w.fx.SettingsViewController - Initializing
09:43:46.351 [JavaFX Application Thread] DEBUG d.p.w.fx.SettingsViewController - Save settings clicked
09:43:46.352 [JavaFX Application Thread] DEBUG d.p.w.fx.SettingsViewController - Closing settings dialogue
09:43:46.352 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - View /fx/views/settings.fxml got close request. Notifying callback
09:43:46.352 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Settings window closed, connection settings set and groupBy combobox empty -> trying to fetch groupByCategories
09:43:46.352 [JavaFX Application Thread] INFO  d.p.w.fx.MainViewController - Fetching GroupByCategories
09:43:46.352 [JavaFX Application Thread] INFO  d.p.w.fx.MainViewController - Starting task GetGroupByCategories-Task
09:43:46.356 [pool-2-thread-1] WARN  d.p.w.y.connector.ApiLoginConnector - Fetching groupBy categories from http://youtrack/rest/reports/timeReports/possibleGroupByCategories failed: Unauthorized
09:43:46.370 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Task GetGroupByCategories-Task changed from READY to SCHEDULED
09:43:46.370 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Task GetGroupByCategories-Task changed from SCHEDULED to RUNNING
09:43:46.372 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Task GetGroupByCategories-Task changed from RUNNING to FAILED
09:43:46.372 [JavaFX Application Thread] WARN  d.p.w.fx.MainViewController - Task GetGroupByCategories-Task failed
09:43:46.373 [JavaFX Application Thread] WARN  d.p.w.fx.MainViewController - Showing error to user
java.lang.IllegalStateException: Could not get categories for groupBy criterias from YouTrack (Message: Unauthorized, StatusCode: 401)
	at de.pbauerochse.worklogviewer.util.ExceptionUtil.getIllegalStateException(ExceptionUtil.java:34) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.util.ExceptionUtil.getIllegalStateException(ExceptionUtil.java:30) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.youtrack.connector.YouTrackConnectorBase.getPossibleGroupByCategories(YouTrackConnectorBase.java:73) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.fx.tasks.GetGroupByCategoriesTask.call(GetGroupByCategoriesTask.java:27) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.fx.tasks.GetGroupByCategoriesTask.call(GetGroupByCategoriesTask.java:14) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at javafx.concurrent.Task$TaskCallable.call(Task.java:1423) ~[jfxrt.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_121]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_121]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_121]
```

**Comment by pbauerochse** [comment]
Did you change the credentials in the settings? In that case it _might_ be necessary to restart the worklog viewer. The response code indicates, that Youtrack is not accepting your username and password combination so I would assume, that the provided credentials might not be correct.

**Comment by m-radzikowski** [comment]
Well, my credentials are correct. When I change them to bad ones I get different exception:

```
1114:02:56.949 [JavaFX Application Thread] DEBUG d.p.w.fx.SettingsViewController - Save settings clicked
14:02:56.950 [JavaFX Application Thread] DEBUG d.p.w.fx.SettingsViewController - Closing settings dialogue
14:02:56.950 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - View /fx/views/settings.fxml got close request. Notifying callback
14:02:56.950 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Settings window closed, connection settings set and groupBy combobox empty -> trying to fetch groupByCategories
14:02:56.950 [JavaFX Application Thread] INFO  d.p.w.fx.MainViewController - Fetching GroupByCategories
14:02:56.950 [JavaFX Application Thread] INFO  d.p.w.fx.MainViewController - Starting task GetGroupByCategories-Task
14:02:56.951 [pool-2-thread-1] DEBUG d.p.w.y.connector.ApiLoginConnector - New connection or connection parameters changed. Performing login
14:02:57.003 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Task GetGroupByCategories-Task changed from READY to SCHEDULED
14:02:57.003 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Task GetGroupByCategories-Task changed from SCHEDULED to RUNNING
14:02:57.004 [JavaFX Application Thread] DEBUG d.p.w.fx.MainViewController - Task GetGroupByCategories-Task changed from RUNNING to FAILED
14:02:57.005 [JavaFX Application Thread] WARN  d.p.w.fx.MainViewController - Task GetGroupByCategories-Task failed
14:02:57.005 [JavaFX Application Thread] WARN  d.p.w.fx.MainViewController - Showing error to user
java.lang.IllegalStateException: Could not login to YouTrack (Message: Forbidden, StatusCode: 403)
	at de.pbauerochse.worklogviewer.util.ExceptionUtil.getIllegalStateException(ExceptionUtil.java:34) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.util.ExceptionUtil.getIllegalStateException(ExceptionUtil.java:30) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.youtrack.connector.ApiLoginConnector.performLoginIfNecessary(ApiLoginConnector.java:59) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.youtrack.connector.YouTrackConnectorBase.getLoggedInClient(YouTrackConnectorBase.java:56) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.youtrack.connector.YouTrackConnectorBase.getPossibleGroupByCategories(YouTrackConnectorBase.java:62) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.fx.tasks.GetGroupByCategoriesTask.call(GetGroupByCategoriesTask.java:27) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at de.pbauerochse.worklogviewer.fx.tasks.GetGroupByCategoriesTask.call(GetGroupByCategoriesTask.java:14) ~[youtrack-worklog-viewer-2.2.3.jar:na]
	at javafx.concurrent.Task$TaskCallable.call(Task.java:1423) ~[jfxrt.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_121]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_121]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_121]
```

I've tried to restart app several times, but it didn't help.

**Comment by pbauerochse** [comment]
Do you get a response when you open the URL (http://youtrack/rest/reports/timeReports/possibleGroupByCategories) in a browser oder do you also get an Exception?

**Comment by m-radzikowski** [comment]
Yes, I'm getting proper XML response (if I'm logged in in youtrack and cookie is attached to request by browser).

When I'm not logged in youtrack I'm getting 401.

**Comment by pbauerochse** [comment]
Could you checkout the sources and debug into line 44 of ApiLoginConnector and check if your password is properly decrypted in the settings object? And if the loginUrl looks correct to you?

**Comment by m-radzikowski** [comment]
Yes, login operation is performed correctly, there is status 200 returned in response.

Ok, I think I've found the reason. We have youtrack on standalone installation inside our network, and the URL to it is simply http://youtrack/. When HttpClient performs login and gets cookies, they are assigned to domain "youtrack". And for some reason they don't work after that for next requests. Weird.

When I changed domain to our alias domain, youtrack.company.local, it works well. Also with IP address instead of domain name it's ok.

So - thank you for your time and help in investigation. It looks like the Apache Http Client has some strange policy with single-domain name cookies, maybe it's somehow justified by security reasons, and maybe it's just a bug.
